File: src/api/decoders/assessment_student_time_accommodation.ts
// Strict decoder for the minimal Instructor accommodation projection.
import type { AssessmentStudentTimeAccommodation } from "../../../generated/api/AssessmentStudentTimeAccommodation";
import {
  DecodeError,
  decodeBoolean,
  decodeFiniteNumber,
  decodeNullable,
  decodePositiveInteger,
  decodeRecord,
  decodeString,
} from "../decoder";

export function decodeStudentTimeMultiplier(value: unknown, path: string): number | null {
  const multiplier = decodeNullable(value, path, decodeFiniteNumber);
  if (multiplier !== null && multiplier < 1) throw new DecodeError(path, "a multiplier >= 1");
  return multiplier;
}

export function decodeAccommodationEditNumber(value: unknown, path: string): string | null {
  const number = decodeNullable(value, path, decodeString);
  if (
    number !== null &&
    (!/^[1-9][0-9]*$/u.test(number) || BigInt(number) > 9_223_372_036_854_775_807n)
  ) {
    throw new DecodeError(path, "a canonical accommodation Edit Number");
  }
  return number;
}

export function decodeAssessmentStudentTimeAccommodation(
  value: unknown,
  path = "response",
): AssessmentStudentTimeAccommodation {
  const record = decodeRecord(value, path);
  const keys = [
    "rosterId",
    "timeMultiplier",
    "editNumber",
    "baseDurationSeconds",
    "effectiveDurationSeconds",
    "cappedAt24Hours",
  ];
  if (
    Object.keys(record).length !== keys.length ||
    Object.keys(record).some((key) => !keys.includes(key))
  ) {
    throw new DecodeError(path, "one closed Student time configuration");
  }
  const rosterId = decodeString(record.rosterId, `${path}.rosterId`);
  if (!/^[A-Za-z0-9._-]{1,64}$/u.test(rosterId)) throw new DecodeError(path, "a Course roster ID");
  const timeMultiplier = decodeStudentTimeMultiplier(
    record.timeMultiplier,
    `${path}.timeMultiplier`,
  );
  const editNumber = decodeAccommodationEditNumber(record.editNumber, `${path}.editNumber`);
  const baseDurationSeconds = decodeNullable(
    record.baseDurationSeconds,
    `${path}.baseDurationSeconds`,
    decodePositiveInteger,
  );
  const effectiveDurationSeconds = decodeNullable(
    record.effectiveDurationSeconds,
    `${path}.effectiveDurationSeconds`,
    decodePositiveInteger,
  );
  const cappedAt24Hours = decodeBoolean(record.cappedAt24Hours, `${path}.cappedAt24Hours`);
  if (
    (baseDurationSeconds === null) !== (effectiveDurationSeconds === null) ||
    (baseDurationSeconds !== null && baseDurationSeconds > 43_200) ||
    (effectiveDurationSeconds !== null &&
      (effectiveDurationSeconds > 86_400 ||
        effectiveDurationSeconds < (baseDurationSeconds ?? 1))) ||
    (editNumber === null && timeMultiplier !== null) ||
    (cappedAt24Hours && effectiveDurationSeconds !== 86_400)
  ) {
    throw new DecodeError(path, "a consistent finite-duration Student time configuration");
  }
  return {
    rosterId,
    timeMultiplier,
    editNumber,
    baseDurationSeconds,
    effectiveDurationSeconds,
    cappedAt24Hours,
  };
}

File: crates/question_model/src/assessment_student_time_accommodation.rs
//! Minimal Course-scoped Instructor Student time configuration wire contract.
use serde::{Deserialize, Serialize};

/// Current time configuration and finite preview for one active Course Student.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AssessmentStudentTimeAccommodation {
    /// Existing Course roster identifier, never a Student Record identity.
    pub roster_id: String,
    /// Unset means standard Assessment time; finite custom values are >= 1.
    pub time_multiplier: Option<f64>,
    /// Current accommodation Edit Number; unset means no accommodation row.
    pub accommodation_edit_number: Option<String>,
    /// Current authored/default base; empty drafts have no duration yet.
    pub base_duration_seconds: Option<u32>,
    /// Multiplied duration rounded upward and capped at 24 hours.
    pub effective_duration_seconds: Option<u32>,
    /// True when the 24-hour ceiling reduces the multiplied duration.
    pub capped_at_24_hours: bool,
}

/// Explicit time-only save, preserving date and Attempt-count adjustments.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct SaveAssessmentStudentTimeAccommodationInput {
    /// Unset clears only the time multiplier.
    pub time_multiplier: Option<f64>,
    /// Current accommodation Edit Number; unset expects no row.
    pub expected_accommodation_edit_number: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn student_time_accommodation_json_uses_accommodation_edit_number() {
        let value = AssessmentStudentTimeAccommodation {
            roster_id: "bio-301".to_string(),
            time_multiplier: None,
            accommodation_edit_number: Some("2".to_string()),
            base_duration_seconds: None,
            effective_duration_seconds: None,
            capped_at_24_hours: false,
        };
        let wire = serde_json::to_value(&value).expect("accommodation serializes");
        assert_eq!(wire["accommodationEditNumber"], "2");
        assert!(wire.get("editNumber").is_none());
        assert!(
            serde_json::from_value::<AssessmentStudentTimeAccommodation>(serde_json::json!({
                "rosterId": "bio-301",
                "timeMultiplier": null,
                "editNumber": "2",
                "baseDurationSeconds": null,
                "effectiveDurationSeconds": null,
                "cappedAt24Hours": false
            }))
            .is_err()
        );
    }
}
